File: RELEASE_ZLS38100_P3_0_0/apps/C/hbi_load_grammar.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include "typedefs.h"
#include "chip.h"
#include "hbi.h"

/* size of buffer for HBI read/write*/
#define MAX_RW_SIZE 64

/* ------------------------------------------------------------ */
void HbiRead(hbi_handle_t handle, reg_addr_t address, user_buffer_t *pData, size_t length) {
    hbi_status_t status;

    status = HBI_read(handle, address, pData, length);
    if(status != HBI_STATUS_SUCCESS) {
        printf("Error - HbiRead(): HBI_read failed (%d)\n", status);
        HBI_close(handle);
        exit(-1);
    }
}

/* ------------------------------------------------------------ */
void HbiWrite(hbi_handle_t handle, reg_addr_t address, user_buffer_t *pData, size_t length) {
    hbi_status_t status;

    status = HBI_write(handle, address, pData, length);
    if(status != HBI_STATUS_SUCCESS) {
        printf("Error - HbiWrite(): HBI_write failed (%d)\n", status);
        HBI_close(handle);
        exit(-1);
    }
}
/* ------------------------------------------------------------ */
unsigned short HbiReadWord(hbi_handle_t handle, reg_addr_t address) {
    user_buffer_t buf[2];

    HbiRead(handle, address, buf, 2);

    return (((unsigned short)buf[0] << 8) + (unsigned short)buf[1]);
}

/* ------------------------------------------------------------ */
void HbiWriteWord(hbi_handle_t handle, reg_addr_t address, unsigned short data) {
    user_buffer_t buf[2];

    buf[0] = (user_buffer_t)(data >> 8);
    buf[1] = (user_buffer_t)(data & 0x00FF);
    HbiWrite(handle, address, buf, 2);
}

/* ------------------------------------------------------------ */
void BusySpinWait(hbi_handle_t handle) {
    while (HbiReadWord(handle, 0x032) == 0xFFFF);
}

/* ------------------------------------------------------------ */
unsigned int Buffer2Int(user_buffer_t *pData) {
    unsigned int integer;

    integer = (unsigned int)pData[0];
    integer = (integer << 8) + (unsigned int)pData[1];
    integer = (integer << 8) + (unsigned int)pData[2];
    integer = (integer << 8) + (unsigned int)pData[3];

    return integer;
}

/* ------------------------------------------------------------ */
void Int2Buffer(unsigned int integer, user_buffer_t *pData) {

    pData[0] = (user_buffer_t)((integer >> 24) & 0x000000FF);
    pData[1] = (user_buffer_t)((integer >> 16) & 0x000000FF);
    pData[2] = (user_buffer_t)((integer >> 8) & 0x000000FF);
    pData[3] = (user_buffer_t)(integer & 0x000000FF);
}

/* ------------------------------------------------------------ */
void CheckHostCmdStatus(hbi_handle_t handle) {
    unsigned short flashStat;

    flashStat = HbiReadWord(handle, 0x034) & 0x000F;
    if (flashStat != 0) {
        switch (flashStat) {
            case 0x1:
                printf("Error - Invalid flash image\n");
                break;

            case 0x2:
                printf("Error - Checksum error during load\n");
                break;

            case 0x3:
                printf("Error - No space in flash for application image size\n");
                break;

            case 0x4:
                printf("Error - Configuration record size doesn’t match application specified\n");
                break;

            case 0x5:
                printf("Error - Invalid flash header data\n");
                break;

            case 0x6:
                printf("Error - No flash present\n");
                break;

            case 0x7:
                printf("Error - Flash failure\n");
                break;

            case 0x8:
                printf("Error - Command failure\n");
                break;

            case 0x9:
                printf("Error - No configuration record available\n");
                break;

            case 0xA:
                printf("Error - Invalid Command - Application running\n");
                break;

            default:
                printf("Error - Unknown flash error\n");
                break;
        }
    }

    if (flashStat != 0) {
        HBI_close(handle);
        exit(-1);
    }
}

/* ------------------------------------------------------------ */
void LoadGrammarFile(hbi_handle_t handle, char * binGrammarPath) {
    size_t byteCount;
    FILE *file;
    user_buffer_t  buf[MAX_RW_SIZE];
    hbi_status_t status;
    user_buffer_t offset, segAddress[4], segAddressTemp[4], segSize[4];
    unsigned short lastSegIndex;
    unsigned int maxSize, grammarSize = 0;

    file = fopen(binGrammarPath, "rb");
    if (file == NULL) {
        printf("Error - LoadGrammarFile(): Cannot open the grammar input file\n");
        HBI_close(handle);
        exit(-1);
    }

    /* Read the ASR segment address */
    HbiRead(handle, 0x4B8, segAddress, 4);
    offset = segAddress[3];

    /* Read the ASR max address */
    HbiRead(handle, 0x4BC, segAddressTemp, 4);

    /* Get the grammar max size */
    maxSize = Buffer2Int(segAddressTemp) - Buffer2Int(segAddress) - 1;

    /* Get the file size */
    fseek(file, 0L, SEEK_END);
    grammarSize = ftell(file);
    fseek(file, 0L, SEEK_SET);
    if (grammarSize > maxSize) {
        printf("Error - LoadGrammarFile(): Grammar file to large (exceeds %d bytes)\n", maxSize);
        HBI_close(handle);
        exit(-1);
    }

    /* Store the start address for later use */
    memcpy(segAddressTemp, segAddress, 4);

    /* Disable the ASR */
    HbiWriteWord(handle, 0x032, 0x800D);
    HbiWriteWord(handle, 0x006, 4);
    BusySpinWait(handle);

    byteCount = MAX_RW_SIZE;
    while (byteCount == MAX_RW_SIZE) {
        /* Read a block of data from the bin file */
        byteCount = fread(buf, 1, MAX_RW_SIZE, file);

        /* Write the address for a page 255 type access */
        HbiWrite(handle, 0x00C, segAddress, 4);

        /* Write the block */
        HbiWrite(handle, 0xFF00 | (reg_addr_t)offset, buf, byteCount);

        /* increment the address pointer */
        Int2Buffer(Buffer2Int(segAddress) + MAX_RW_SIZE, segAddress);
        offset = segAddress[3];
    }

    fclose(file);

    /* Update the segment table */
    /* Recover the start address */
    memcpy(segAddress, segAddressTemp, 4);

    /* get the number of segments */
    lastSegIndex = HbiReadWord(handle, 0x13E) - 1;

    /* Read the load address of the last segment */
    HbiRead(handle, 0x144 + 8 * lastSegIndex, segAddressTemp, 4);

    /* Convert the grammar size in a buffer */
    Int2Buffer(grammarSize, segSize);

    /* If the last segment address is an ASR segment, update the size otherwise create a new segment */
    if (Buffer2Int(segAddressTemp) == Buffer2Int(segAddress)) {
        /* Update the last segment size */
        HbiWrite(handle, 0x140 + 8 * lastSegIndex, segSize, 4);
    } else {
        /* Create a new segment */
        lastSegIndex++;
        HbiWrite(handle, 0x140 + 8 * lastSegIndex, segSize, 4);
        HbiWrite(handle, 0x144 + 8 * lastSegIndex, segAddress, 4);
        HbiWriteWord(handle, 0x13E, lastSegIndex + 1);
    }

    /* Enable the ASR */
    HbiWriteWord(handle, 0x032, 0x800E);
    HbiWriteWord(handle, 0x006, 4);
    BusySpinWait(handle);

    printf("Info - Grammar successfully loaded to RAM\n");
}

/* ------------------------------------------------------------ */
void printHelp(char * exeName) {
    printf("\n-= ASR Grammar Manager =-\n");
    printf("Usage: %s -[d deviceId] [-q] [-l binPath] [-f] [-s flashSlot] [-h]\n", exeName);
    printf("    - \"-q\":\n");
    printf("        Query the number of grammars stored in flash\n");
    printf("    - \"-l binPath\":\n");
    printf("        Load a grammar from a bin file to RAM\n");
    printf("        Note: generate the bin file using \"tw_grammar_converter.py\"\n");
    printf("    - \"-f\":\n");
    printf("        Save the grammar from RAM to flash\n");
    printf("    - \"-s flashSlot\":\n");
    printf("        Swap grammars from flash\n");
    printf("    - \"-h\":\n");
    printf("        Prints that help\n\n");
}

/* ------------------------------------------------------------ */
int main(int argc, char** argv) {
    int c, numGrammars, queryNumGrammar = 0, saveToFlash = 0, grammarIdx = -1;
    char *binGrammarPath = NULL;
    hbi_status_t status;
    hbi_dev_cfg_t devConfig;
    hbi_handle_t handle;
    hbi_device_id_t deviceId = -1;

    /* Optional arguments */
    while((c = getopt(argc,argv,"hql:fs:d:")) != -1) {
        switch(c) {
            case 'd':
                deviceId= strtol(optarg,NULL,16);
                break;
            case 'h':
                printHelp(argv[0]);
                return 0;
                break;

            case 'q':
                /* Query the number of grammars stored in flash */
                queryNumGrammar = 1;
                break;

            case 'l':
                /* Load a grammar from a bin file to RAM */
                binGrammarPath = optarg;
                break;

            case 'f':
                /* Save the grammar from RAM to flash */
                saveToFlash = 1;
                break;

            case 's':
                /* Swap grammars from flash */
                grammarIdx = atoi(optarg);
                break;

            case '?':
                return -1;
                break;

            default:
                abort();
        }
    }

    /* No arguments */
    if (argc == 1) {
        printHelp(argv[0]);
        return 0;
    }

    /* Open the HBI driver */
    devConfig.deviceId = deviceId;
    devConfig.pDevName = NULL;


    status = HBI_open(&handle, &devConfig);
    if(status != HBI_STATUS_SUCCESS) {
        printf("Error - Main(): HBI_open failed (%d)\n", status);
        return -1;
    }

    /* The firmware needs to be running in order to manage grammars */
    if ((HbiReadWord(handle, 0x028) & 0x8000) == 0) {
        printf("Error - Main(): Application firmware stopped\n");
        HBI_close(handle);
        exit(-1);
    }

    /* Get the number of images in flash */
    numGrammars = HbiReadWord(handle, 0x04E);

    if (queryNumGrammar) {
        printf("Info - Number of grammars in flash: %d\n", numGrammars);
    } else if (binGrammarPath != NULL) {
        LoadGrammarFile(handle, binGrammarPath);
        if (saveToFlash) {
            /* Save the grammar from RAM to the next flash slot */
            HbiWriteWord(handle, 0x034, numGrammars + 1);
            HbiWriteWord(handle, 0x032, 0x800B);
            HbiWriteWord(handle, 0x006, 4);
            BusySpinWait(handle);
            CheckHostCmdStatus(handle);
            printf("Info - Grammar successfully saved to flash slot: %d\n", numGrammars + 1);
        }
    } else if (grammarIdx != -1) {
        if ((grammarIdx == 0) || (grammarIdx > numGrammars)) {
            printf("Error - Illegal flash index, should be: 1 <= N <= %d\n", numGrammars);
        } else {
            /* Load the specified "grammarIdx" from flash */
            HbiWriteWord(handle, 0x034, grammarIdx);
            HbiWriteWord(handle, 0x032, 0x800C);
            HbiWriteWord(handle, 0x006, 4);
            BusySpinWait(handle);
            CheckHostCmdStatus(handle);
            printf("Info - Grammar successfully loaded from flash slot: %d\n", grammarIdx);
        }
    }

    /* Close the HBI driver */
    HBI_close(handle);

    return 0;
}

File: RELEASE_ZLS38100_P3_0_0/apps/Python/hbi_load_grammar.py
#!/usr/bin/env python

# Functions designed for programmatic interface:
# grammar = CombineGrammar(trigger_am_path, trigger_so_path, command_am_path, command_so_path, description)
#     Returns a buffer containing the combined grammar
#     Use the keyword “None” when an input path is missing, description is mandatory
#
# CheckGrammarSize(handle, grammar_size)
#     Generates an error if the grammar size exceeds the size specified in the configuration record
#
# LoadGrammar(handle, grammar)
#     Loads the previously combined grammar in the Timberwolf memory
#
# grm_flash_slot = SaveGrammarToFlash(handle)
#     Saves the grammar from RAM to FLASH and returns the grammar flash index

from os.path import dirname, realpath, isfile
import sys
import struct
import argparse
sys.path.append(dirname(realpath(__file__)) + "/../../../libs")
from hbi import *

# Globals
programmatic = True

# ****************************************************************************
def FormatNumber(res_list):
    number = 0

    for byteNum in res_list:
        number = (number << 8) + byteNum

    return number

# ****************************************************************************
def BusySpinWait(handle):
    # The register 0x032 reads 0xFFFF while the Timberwolf is processing a command
    while (FormatNumber(HBI_read(handle, 0x032, 2)) == 0xFFFF):
        pass

# ****************************************************************************
def LoadGrammar(handle, grammar):
    global programmatic

    # Turn off the ASR
    HBI_write(handle, 0x032, (0x80, 0x0D))
    HBI_write(handle, 0x006, (0x00, 0x04))
    BusySpinWait(handle)

    # Grammar start address
    start_add = FormatNumber(HBI_read(handle, 0x4B8, 4))

    # Copy the grammar in memory (64 bytes block or less)
    grammar_size = len(grammar)
    block_num = grammar_size // 64
    block_rem = grammar_size % 64
    if (block_rem != 0):
        block_num += 1

    for idx in xrange(0, block_num):
        address = start_add + idx * 64
        address_seq = struct.unpack("4B", struct.pack(">I", address))
        hbi_offset = address_seq[3]
        HBI_write(handle, 0x00C, address_seq)

        if ((idx == (block_num - 1)) and (block_rem != 0)):
            block_size = block_rem
        else:
            block_size = 64
        grammar_block = struct.unpack(str(block_size) + "B", grammar[idx * 64:idx * 64 + block_size])
        HBI_write(handle, 0xFF00 + hbi_offset, grammar_block)

    # Update the segment table
    idx_last_segment = FormatNumber(HBI_read(handle, 0x13E, 2)) - 1

    # Check if the last segment is already an ASR segment (if so, start_add should match that segment address)
    last_segment_add = FormatNumber(HBI_read(handle, 0x144 + 8 * idx_last_segment, 4))
    grammar_size_seq = struct.unpack("4B", struct.pack(">I", grammar_size))
    if (start_add == last_segment_add):
        # Only update the segment size
        HBI_write(handle, 0x140 + 8 * idx_last_segment, grammar_size_seq)
    else:
        # Add a new segment to the table for the ASR and update the segment counter
        if not programmatic:
            print "Info - Create a new segment"
        idx_last_segment += 1
        last_segment_add_seq = struct.unpack("4B", struct.pack(">I", start_add))
        HBI_write(handle, 0x144 + 8 * idx_last_segment, last_segment_add_seq)
        HBI_write(handle, 0x140 + 8 * idx_last_segment, grammar_size_seq)
        HBI_write(handle, 0x13E, (0x00, idx_last_segment + 1))

    # Start the ASR
    HBI_write(handle, 0x032, (0x80, 0x0E))
    HBI_write(handle, 0x006, (0x00, 0x04))
    BusySpinWait(handle)

    if not programmatic:
        print "Info - Grammar successfully loaded to RAM"

# ****************************************************************************
def SaveGrammarToFlash(handle):
    # Read the number of grammars in flash
    images_flash = FormatNumber(HBI_read(handle, 0x04E, 2))

    # Save the current image in the next slot (slot indexes start at 1)
    HBI_write(handle, 0x032, (0x80, 0x0B))
    HBI_write(handle, 0x034, (0x00, images_flash + 1))
    HBI_write(handle, 0x006, (0x00, 0x04))
    BusySpinWait(handle)

    # Check the flash operation status
    status = FormatNumber(HBI_read(handle, 0x034, 2)) & 0x000F
    if (status != 0):
        raise ValueError("Error - SaveGrammarToFlash(): Grammar could not be saved to flash (error 0x%X)" % status)

    return images_flash + 1

# ****************************************************************************
def CheckGrammarSize(handle, grammar_size):
    start_add = FormatNumber(HBI_read(handle, 0x4B8, 4))
    endAdd = FormatNumber(HBI_read(handle, 0x4BC, 4))
    if (grammar_size > (endAdd - start_add + 1)):
        raise ValueError("Error - CheckGrammarSize(): The grammar exceeds the memory size")

# ****************************************************************************
def CreateHeader(trigger_so_size, trigger_am_size, command_so_size, command_am_size, description, version):
#typedef struct
#{
#   const u16 *trigger_grammar_ptr;
#   const u16 *trigger_net_ptr;
#   const u16 *command_grammar_ptr;
#   const u16 *command_net_ptr;
#   char  description[32];
#   u16 version;
#   char  rsrvd[14];
# } GRAMMAR_DESCRIPTOR;

    desc_bytes = bytearray(64)

    # Note: "!=" is effectively an XOR
    if (trigger_so_size == 0) != (trigger_am_size == 0):
        raise ValueError("Error - CreateHeader(): Part of the Trigger grammar is missing")

    if (command_so_size == 0) != (command_am_size == 0):
        raise ValueError("Error - CreateHeader(): Part of the Command grammar is missing")

    if (trigger_so_size == 0) and (trigger_am_size == 0) and (command_so_size == 0) and (command_am_size == 0):
        raise ValueError("Error - CreateHeader(): No grammar to load")

    trigger_so_offset = 0
    trigger_am_offset = 0
    command_so_offset = 0
    command_am_offset = 0

    # Note: The offset of 0x80 is to store two copies of the GRAMMAR_DESCRIPTOR structure
    if (trigger_so_size != 0):
        trigger_so_offset = 0x80
        trigger_am_offset = trigger_so_offset + trigger_so_size
    else:
        command_so_offset = 0x80

    if (command_so_size != 0):
        # Check if a Trigger grammar exists
        if (command_so_offset != 0x80):
            command_so_offset = trigger_am_offset + trigger_am_size
        command_am_offset = command_so_offset + command_so_size

    struct.pack_into(">IIII", desc_bytes, 0, trigger_so_offset, trigger_am_offset, command_so_offset, command_am_offset)

    # Store the grammar description (max 32 char)
    gram_desc_bytes = bytes(description)
    endindex = len(gram_desc_bytes)
    if(endindex >= 32):
        endindex = 31
    endindex = endindex + 16
    desc_bytes[16:endindex] = gram_desc_bytes;

    # Store the version
    struct.pack_into(">H", desc_bytes, 48, version)

    return desc_bytes

# ****************************************************************************
def ParseGrammarBinFile(path):
    # Note: "b" in "rb" is only required for Windows to read in binary but doesn't hurt on Linux
    f = open(path, "rb")
    buf = f.read()
    f.close()
    return buf, len(buf)

# ****************************************************************************
def CombineGrammar(trigger_am_path, trigger_so_path, command_am_path, command_so_path, description):
    global programmatic

    # Parse all the grammar files
    if ((trigger_am_path != None) and isfile(trigger_am_path)):
        trigger_am_buf, trigger_am_size = ParseGrammarBinFile(trigger_am_path)
    else:
        if not programmatic:
            print "Info - No Trigger acoustic model"
        trigger_am_buf = ""
        trigger_am_size = 0

    if ((trigger_so_path != None) and isfile(trigger_so_path)):
        trigger_so_buf, trigger_so_size = ParseGrammarBinFile(trigger_so_path)
    else:
        if not programmatic:
            print "Info - No Trigger search object"
        trigger_so_buf = ""
        trigger_so_size = 0

    if ((command_am_path != None) and isfile(command_am_path)):
        command_am_buf, command_am_size = ParseGrammarBinFile(command_am_path)
    else:
        if not programmatic:
            print "Info - No Command acoustic model"
        command_am_buf = ""
        command_am_size = 0

    if ((command_so_path != None) and isfile(command_so_path)):
        command_so_buf, command_so_size = ParseGrammarBinFile(command_so_path)
    else:
        if not programmatic:
            print "Info - No Command search object"
        command_so_buf = ""
        command_so_size = 0

    # Create the description stucture (version 1)
    structure = CreateHeader(trigger_so_size, trigger_am_size, command_so_size, command_am_size, description, 1)

    # Merge all the pieces of the grammar inluding two copies of the description structure
    return structure + structure + trigger_so_buf + trigger_am_buf + command_so_buf + command_am_buf

# ****************************************************************************
if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        formatter_class = argparse.RawDescriptionHelpFormatter,
        description = "Timberwolf ASR Grammar loader",
        epilog = ("""
This tool combines all the pieces (bin files) of a Sensory (TM)
Automatic Speech Recognition (ASR) model and loads them in
an ASR enabled Timberwolf device

ex: Load a full grammar with a trigger word and cammand set then save it to flash
    %s -tam trigger_am.bin -tso trigger_so.bin -cam command_am.bin -cso command_so -d "my full grammar" -s
ex: Load a trigger only grammar then save it to flash
    %s -tam trigger_am.bin -tso trigger_so.bin -d "my trigger only grammar" -s
""" % (sys.argv[0], sys.argv[0])))
    parser.add_argument("-tam", "--triggerAM", help = "trigger acoustic model (*.bin)")
    parser.add_argument("-tso", "--triggerSO", help = "trigger search object (*.bin)")
    parser.add_argument("-cam", "--commandAM", help = "command acoustic model (*.bin)")
    parser.add_argument("-cso", "--commandSO", help = "command search object (*.bin)")
    parser.add_argument("-d", "--description", help = "grammar decription string (up to 32 characters)", default = "")
    parser.add_argument("-s", "--saveToFlash", help = "save the grammar to flash", action = "store_true")

    # Print the help if no arguments are passed
    if (len(sys.argv[1:]) == 0):
        parser.print_help()
        parser.exit()

    # Parse the input arguments
    args = parser.parse_args()

    programmatic = False

    try:
        grammar = CombineGrammar(args.triggerAM, args.triggerSO, args.commandAM, args.commandSO, args.description)

    except ValueError as err:
        print err
        sys.exit()

    # Init the HBI driver
    cfg = hbi_dev_cfg_t();
    handle = HBI_open(cfg)

    try:
        # Check if the grammar fits in the Timberwolf
        CheckGrammarSize(handle, len(grammar))

        # Load the grammar in memory
        LoadGrammar(handle, grammar)

        if args.saveToFlash:
            slot = SaveGrammarToFlash(handle)
            print "Info - Grammar successfully saved to flash slot %d" % slot

    except ValueError as err:
        print err

    # Close HBI driver
    HBI_close(handle)
